grayjay: init at 5

samfundev opened the pull request:
Adds Grayjay Desktop. Fixes #366543

This is the first time I've attempted to package an application so I appreciate any help. I tested this by opening the application, installing the YouTube source and playing a video. I copied the runtime dependencies from [here](https://gitlab.futo.org/videostreaming/Grayjay.Desktop#nixos-config), but I'm not sure if all of those are necessary. Especially since the updater is something that I intentionally left broken. I should also give credit to @Electrostasy whose [nix file](https://github.com/NixOS/nixpkgs/issues/366543#issuecomment-2559966100) I built off of.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

samfundev committed:
maintainers: add samfundev

ddogfoodd reviewed (commented):
builds and opens for me on x86_64-linux

but somehow can't sync devices or install sources, might be just the alpha state or my network idk

maybe some other person can test and approve

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+  version = "3";
+  src = fetchgit {
+    url = "https://gitlab.futo.org/videostreaming/Grayjay.Desktop";
+    rev = "refs/tags/${version}";
```
```suggestion
    tag = version;
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+    npmDepsHash = "sha256-pTEbMSAJwTY6ZRriPWfBFnRHSYufSsD0d+hWGz35xFM=";
+
+    installPhase = ''
+      cp -r dist/ $out
```
Nit:


```suggestion
      runHook preInstall
      cp -r dist/ $out
      runHook postInstall
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+in
+buildDotnetModule {
+  pname = "grayjay";
+  version = "3";
+
+  inherit src;
```
```suggestion
  inherit version src;
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+
+  executables = [ "Grayjay" ];
+
+  preBuild = ''
+    rm -r Grayjay.ClientServer/wwwroot/web
+    cp -r ${frontend} Grayjay.ClientServer/wwwroot/web'';
```
```suggestion
  preBuild = ''
    rm -r Grayjay.ClientServer/wwwroot/web
    cp -r ${frontend} Grayjay.ClientServer/wwwroot/web
  ''; 
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+    libsecret
+  ];
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

Pandapip1 reviewed (changes requested):
First pass

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+    longDescription = ''Grayjay is a multi-platform media application that allows you to watch content from multiple platforms in a single application. Using an extendable plugin system developers can make new integrations with additional platforms. Plugins are cross-compatible between Android and Desktop.'';
+    homepage = "https://grayjay.app/desktop/";
+    license = {
+      fullName = "FUTO Source First License 1.1";
+      url = "https://gitlab.futo.org/videostreaming/Grayjay.Desktop/-/blob/master/LICENSE.md";
+    };
```
Should be added to the license list. Also missing `free = false` and `redistributable = true`.

https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+    name = "grayjay-frontend";
+    inherit version;
+
+    src = "${src}/Grayjay.Desktop.Web";
```
Why not use `sourceRoot`?

```suggestion
    inherit src;

    sourceRoot = "Grayjay.Desktop.Web";
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,117 @@
+    "${placeholder "out"}/lib/grayjay"
+  ];
+
+  runtimeDeps = with pkgs; [
```
Each of these should be individually brought in instead of using `with pkgs` so that the inputs can be overridden.

Pandapip1 commented on lib/licenses.nix:
```diff
@@ -1139,6 +1139,13 @@ lib.mapAttrs mkLicense ({
+    fullName = "Source First License 1.1";
+    url = "https://gitlab.futo.org/videostreaming/grayjay/-/blob/master/LICENSE.md";
+    free = false;
+    redistributable = false;
```
```suggestion
    redistributable = true;
```

> You may distribute the software or provide it to others only if you do so free of charge for non-commercial purposes.

Pandapip1 replied:
Feel free to mark resolved comments as resolved!

Pandapip1 commented on lib/licenses.nix:
Nit: Adding the license should probably be its own individual commit.

samfundev commented:
Thanks for the help! This is the first time I've attempted to package an application so I appreciate it. I tested this by opening the application, installing the YouTube source and playing a video. I copied the runtime dependencies from [here](https://gitlab.futo.org/videostreaming/Grayjay.Desktop#nixos-config), but I'm not sure if all of those are necessary. Especially since the updater is something that I intentionally left broken. I should also give credit to @Electrostasy whose [nix file](https://github.com/NixOS/nixpkgs/issues/366543#issuecomment-2559966100) I built off of.

samfundev committed:
lib/licenses: add sfl

samfundev committed:
grayjay: init at version 5

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,136 @@
+  version = "3";
+  src = fetchgit {
+    url = "https://gitlab.futo.org/videostreaming/Grayjay.Desktop";
+    rev = version;
```
```suggestion
    tag = version;
```

Pandapip1 reviewed (changes requested):
I just tried building this. It looks like you've gotten bit by the fact that fixed-output derivations are cached, since right now the fetcher is failing:

```log
error: builder for '/nix/store/skv4a9hircqwm1gz9n5hbix9yrqf28mc-Grayjay.Desktop-3.drv' failed with exit code 1;
       last 16 log lines:
       > exporting https://gitlab.futo.org/videostreaming/Grayjay.Desktop (rev 3) into /nix/store/wx65wpm8l63clbv2l0v141kxfrpz8ahf-Grayjay.Desktop-3
       > Git LFS initialized.
       > Initialized empty Git repository in /nix/store/wx65wpm8l63clbv2l0v141kxfrpz8ahf-Grayjay.Desktop-3/.git/
       > warning: redirecting to https://gitlab.futo.org/videostreaming/Grayjay.Desktop.git/
       > warning: redirecting to https://gitlab.futo.org/videostreaming/Grayjay.Desktop.git/
       > remote: Enumerating objects: 1051, done.
       > remote: Counting objects: 100% (1051/1051), done.
       > remote: Compressing objects: 100% (881/881), done.
       > remote: Total 1051 (delta 124), reused 1037 (delta 120), pack-reused 0 (from 0)
       > Receiving objects: 100% (1051/1051), 146.37 MiB | 6.77 MiB/s, done.
       > Resolving deltas: 100% (124/124), done.
       > From https://gitlab.futo.org/videostreaming/Grayjay.Desktop
       >  * tag               3          -> FETCH_HEAD
       > fatal: Not a valid object name 3
       > Unrecognized git object type:
       > Unable to checkout 3 from https://gitlab.futo.org/videostreaming/Grayjay.Desktop.
       For full logs, run 'nix log /nix/store/skv4a9hircqwm1gz9n5hbix9yrqf28mc-Grayjay.Desktop-3.drv'.
error: 1 dependencies of derivation '/nix/store/z590gqz5gbnlrdznvy2vcxigm5dlzcfa-grayjay-3.drv' failed to build
```

Pandapip1 commented:
Looks like the fetcher is currently failing, for some reason. I'm bumping into https://github.com/git-lfs/git-lfs/issues/5511: https://gitlab.futo.org/videostreaming/JustCef/-/tree/master/prebuilt?ref_type=heads

```log
Grayjay.Desktop> remote: Enumerating objects: 38, done.
Grayjay.Desktop> remote: Counting objects: 100% (38/38), done.
Grayjay.Desktop> remote: Compressing objects: 100% (20/20), done.
Grayjay.Desktop> remote: Total 20 (delta 15), reused 5 (delta 0), pack-reused 0 (from 0)
Grayjay.Desktop> Unpacking objects: 100% (20/20), 1.99 KiB | 290.00 KiB/s, done.
Grayjay.Desktop> From https://gitlab.futo.org/videostreaming/JustCef
Grayjay.Desktop>  * branch            56f844fee7b8d26becbf94d4701e40ba7cd280d5 -> FETCH_HEAD
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
```

samfundev commented:
While I do see those lines when it's fetching, I'm still able to build the package. Maybe I'm misunderstanding what you're trying to say.

Pandapip1 commented:
> While I do see those lines when it's fetching, I'm still able to build the package. Maybe I'm misunderstanding what you're trying to say.

Weird, the fetcher hangs for me.

alyssais commented on lib/licenses.nix:
```diff
@@ -1139,6 +1139,13 @@ lib.mapAttrs mkLicense ({
 
+  sfl = {
+    fullName = "Source First License 1.1";
+    url = "https://gitlab.futo.org/videostreaming/grayjay/-/blob/master/LICENSE.md";
```
Is there no more canonical URL?  If not, maybe would make more sense to just inline this since it seems unlikely to be used for anything else.

Pandapip1 replied:
I suggested that this be added to licenses.nix; given what FUTO seems to have planned, I would be surprised if this remains their only SFL-licensed program suitable for inclusion in NixOS.

alyssais replied:
Okay, that's fine.

Pandapip1 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

gepbird commented:
I've got the same seemingly never finishing fetcher output as https://github.com/NixOS/nixpkgs/pull/368427#issuecomment-2564398506

```
Grayjay.Desktop> Cloning into '/nix/store/ljmnynh86lay51kg3y1wkq5x4zg2nzar-Grayjay.Desktop/JustCef'...
Grayjay.Desktop> remote: Enumerating objects: 575, done.        
Grayjay.Desktop> remote: Counting objects: 100% (575/575), done.        
Grayjay.Desktop> remote: Compressing objects: 100% (482/482), done.        
Grayjay.Desktop> remote: Total 575 (delta 8), reused 549 (delta 8), pack-reused 0 (from 0)        
Grayjay.Desktop> Receiving objects: 100% (575/575), 355.42 KiB | 677.00 KiB/s, done.
Grayjay.Desktop> Resolving deltas: 100% (8/8), done.
Grayjay.Desktop> Submodule path 'FUTO.MDNS': checked out '3c7ccbd2df32d454d7b06a0933d1944e9fbc48fa'
Grayjay.Desktop> Submodule path 'Grayjay.Engine': checked out '130bdd8b4a7f16cacb6785f34869890abdc16079'
Grayjay.Desktop> remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
Grayjay.Desktop> remote: Enumerating objects: 38, done.
Grayjay.Desktop> remote: Counting objects: 100% (38/38), done.
Grayjay.Desktop> remote: Compressing objects: 100% (20/20), done.
Grayjay.Desktop> remote: Total 20 (delta 15), reused 5 (delta 0), pack-reused 0 (from 0)
Grayjay.Desktop> Unpacking objects: 100% (20/20), 2.07 KiB | 68.00 KiB/s, done.
Grayjay.Desktop> From https://gitlab.futo.org/videostreaming/JustCef
Grayjay.Desktop>  * branch            56f844fee7b8d26becbf94d4701e40ba7cd280d5 -> FETCH_HEAD
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
Grayjay.Desktop> sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
Grayjay.Desktop> Submodule path 'JustCef': checked out '56f844fee7b8d26becbf94d4701e40ba7cd280d5'
Grayjay.Desktop> removing `.git'...
grayjay-frontend-npm-deps> building '/nix/store/slsi3yb3b3914cpixdx33j3660rxh5r2-grayjay-frontend-npm-deps.drv'
```

But surprisingly after an hour it finishes

![2025-01-01_17-29](https://github.com/user-attachments/assets/a1dc8074-f586-4b47-92b6-87d6ec1d2d93)

Pandapip1 commented:
Well, it builds. I also ran it and it seems to run fine. So I guess I'll take another look at the derivation?

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,136 @@
+    libGL
+    libsecret
+  ];
+
```
Might be worth adding `nix-update-script`. See https://wiki.nixos.org/wiki/Nixpkgs/Update_Scripts

samfundev replied:
nix-update-script doesn't seem to work with fetchgit, so I added gitUpdater instead.

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,136 @@
+    libGL
+    libsecret
+  ];
+
```
Might be worth adding `versionCheckHook`, if GrayJay desktop has a help or version command. See https://nixos.org/manual/nixpkgs/unstable/#versioncheckhook

samfundev replied:
Grayjay doesn't seem to have help or version. There is a file called AppVersion.json but that doesn't seem to have been updated for version 3 because it still says version 2.

Pandapip1 replied:
The point of it is to ensure that the program runs. Feel free to disregard this.

Pandapip1 reviewed (approved):
LGTM, just two nits

samfundev commented:
I think the reason that LFS is taking so long is because it's downloading prebuilt binaries in JustCef. I'm currently using the prebuilt binaries, but perhaps we could skip LFS all together if JustCef was included as part of the build. I could try looking into building that instead of using the prebuilt binaries, but it looks somewhat complicated.

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,139 @@
+}:
+let
+  version = "3";
+  src = fetchgit {
+    url = "https://gitlab.futo.org/videostreaming/Grayjay.Desktop";
+    tag = version;
+    hash = "sha256-gI1M/rDTSH16eeknHJfLA2iawoA5iGTKHfwUW4/lvv8=";
+    fetchSubmodules = true;
+    fetchLFS = true;
+  };
```
```suggestion
  src = fetchFromGitLab {
    domain = "gitlab.futo.org";
    owner = "videostreaming";
    repo = "Grayjay.Desktop";
    tag = version;
    hash = "sha256-gI1M/rDTSH16eeknHJfLA2iawoA5iGTKHfwUW4/lvv8=";
    fetchSubmodules = true;
    fetchLFS = true;
  };
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,139 @@
+
+  meta = {
+    description = "Multi-platform media application that allows you to watch content from multiple platforms in a single application";
+    longDescription = ''Grayjay is a multi-platform media application that allows you to watch content from multiple platforms in a single application. Using an extendable plugin system developers can make new integrations with additional platforms. Plugins are cross-compatible between Android and Desktop.'';
```
```suggestion
    longDescription = "Grayjay is a multi-platform media application that allows you to watch content from multiple platforms in a single application. Using an extendable plugin system developers can make new integrations with additional platforms. Plugins are cross-compatible between Android and Desktop.";
```

Pandapip1 commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,139 @@
+    homepage = "https://grayjay.app/desktop/";
+    license = lib.licenses.sfl;
+    maintainers = with lib.maintainers; [ samfundev ];
+    platforms = lib.platforms.linux;
```
```suggestion
    platforms = lib.platforms.unix;
```

This allows it to be run on the BSDs.

lzszt commented:
The build works for me, but I get the following error when running grayjay.
<details>
<summary>Log output</summary>
```
CEF:Searching for dotcefnative, search paths:
CEF: - /nix/store/q1gw4lh12i35c6z9m2dxmz43782asmx5-grayjay-3/lib/grayjay/cef/dotcefnative
CEF: - /nix/store/q1gw4lh12i35c6z9m2dxmz43782asmx5-grayjay-3/lib/grayjay/dotcefnative
CEF:Working directory '/nix/store/q1gw4lh12i35c6z9m2dxmz43782asmx5-grayjay-3/lib/grayjay/cef'.
CEF:CEF exe path '/nix/store/q1gw4lh12i35c6z9m2dxmz43782asmx5-grayjay-3/lib/grayjay/cef/dotcefnative'.
--change-stack-guard-on-fork=disable --disable-chrome-runtime --parent-to-child 115 --child-to-parent 118 --disable-web-security --use-views 
Program:Main: Starting DotCefProcess finished (10ms)
Program:Main: Starting window.
Receive loop started.
CEF:Could not start dynamically linked executable: /nix/store/q1gw4lh12i35c6z9m2dxmz43782asmx5-grayjay-3/lib/grayjay/cef/dotcefnative
CEF:NixOS cannot run dynamically linked executables intended for generic
CEF:linux environments out of the box. For more information, see:
CEF:https://nix.dev/permalink/stub-ld
GrayjayServer:RunServerAsync: Called with (proxyUrl = ).
Receive loop stopped.
Unhandled exception. System.AggregateException: One or more errors occurred. (Broken pipe)
 ---> System.IO.IOException: Broken pipe
 ---> System.Net.Sockets.SocketException (32): Broken pipe
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.CreateException(SocketError error, Boolean forAsyncThrow)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.SendAsync(Socket socket, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.SendAsync(ReadOnlyMemory`1 buffer, SocketFlags socketFlags, CancellationToken cancellationToken)
   at System.IO.Pipes.PipeStream.WriteAsyncCore(ReadOnlyMemory`1 source, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.IO.Pipes.PipeStream.WriteAsyncCore(ReadOnlyMemory`1 source, CancellationToken cancellationToken)
   at System.IO.Pipes.PipeStream.WriteAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at DotCef.DotCefProcess.CallAsync(OpcodeController opcode, Byte[] body, Int32 offset, Int32 size, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at DotCef.DotCefProcess.CallAsync(OpcodeController opcode, Byte[] body, Int32 offset, Int32 size, CancellationToken cancellationToken)
   at DotCef.DotCefProcess.CreateWindowAsync(String url, Int32 minimumWidth, Int32 minimumHeight, Int32 preferredWidth, Int32 preferredHeight, Boolean fullscreen, Boolean contextMenuEnable, Boolean shown, Boolean developerToolsEnabled, Boolean resizable, Boolean frameless, Boolean centered, Boolean proxyRequests, Boolean logConsole, Func`3 requestProxy, Boolean modifyRequests, Func`3 requestModifier, Boolean modifyRequestBody, String title, String iconPath, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at DotCef.DotCefProcess.CreateWindowAsync(String url, Int32 minimumWidth, Int32 minimumHeight, Int32 preferredWidth, Int32 preferredHeight, Boolean fullscreen, Boolean contextMenuEnable, Boolean shown, Boolean developerToolsEnabled, Boolean resizable, Boolean frameless, Boolean centered, Boolean proxyRequests, Boolean logConsole, Func`3 requestProxy, Boolean modifyRequests, Func`3 requestModifier, Boolean modifyRequestBody, String title, String iconPath, CancellationToken cancellationToken)
   at Grayjay.Desktop.Program.Main(String[] args)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Grayjay.Desktop.Program.Main(String[] args)
   at Grayjay.Desktop.Program.<Main>(String[] args)
--- End of stack trace from previous location ---
   at System.IO.Pipes.PipeStream.WriteAsyncCore(ReadOnlyMemory`1 source, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.IO.Pipes.PipeStream.WriteAsyncCore(ReadOnlyMemory`1 source, CancellationToken cancellationToken)
   at DotCef.DotCefProcess.CallAsync(OpcodeController opcode, Byte[] body, Int32 offset, Int32 size, CancellationToken cancellationToken)
   at DotCef.DotCefProcess.CreateWindowAsync(String url, Int32 minimumWidth, Int32 minimumHeight, Int32 preferredWidth, Int32 preferredHeight, Boolean fullscreen, Boolean contextMenuEnable, Boolean shown, Boolean developerToolsEnabled, Boolean resizable, Boolean frameless, Boolean centered, Boolean proxyRequests, Boolean logConsole, Func`3 requestProxy, Boolean modifyRequests, Func`3 requestModifier, Boolean modifyRequestBody, String title, String iconPath, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Grayjay.Desktop.Program.Main(String[] args)
   at Grayjay.Desktop.Program.<Main>(String[] args)
   ```
</details>

samfundev commented:
Sorry, I had nix-ld enabled. I've added autoPatchelfHook and moved the dependencies to buildInputs. Also tested playing and downloading a video again.

normalcea commented:
I tested this PR on my non-NixOS Linux Mint install with X11 Cinnamon and NixGL. After adding a YouTube source, no audio would play and the desktop detected no audio playing from grayjay (didn't show up at all)

``` bash

# Likely nixGL specific on my end

CEF:MESA-LOADER: failed to open dri: /nix/store/ak5hicfnzhd1kwk7dzlyy4w8sg8zq8hn-mesa-24.3.3/lib/gbm/dri_gbm.so: cannot open shared object file: No such file or directory (search paths /run/opengl-driver/lib/gbm:/nix/store/ak5hicfnzhd1kwk7dzlyy4w8sg8zq8hn-mesa-24.3.3/lib/gbm, suffix _gbm)

# (later when playing a video, no audio plays)

CEF:ALSA lib dlmisc.c:339:(snd_dlobj_cache_get0) Cannot open shared library libasound_module_pcm_pipewire.so (/nix/store/y26wywsahsr3swpnlm4b41px4c035mva-alsa-lib-1.2.13/lib/alsa-lib/libasound_module_pcm_pipewire.so: cannot open shared object file: No such file or directory)
CEF:[2761966:2761966:0119/164821.833698:ERROR:alsa_util.cc(204)] PcmOpen: default,No such device or address
CEF:ALSA lib dlmisc.c:339:(snd_dlobj_cache_get0) Cannot open shared library libasound_module_pcm_pipewire.so (/nix/store/y26wywsahsr3swpnlm4b41px4c035mva-alsa-lib-1.2.13/lib/alsa-lib/libasound_module_pcm_pipewire.so: cannot open shared object file: No such file or directory)
CEF:[2761966:2761966:0119/164821.833781:ERROR:alsa_util.cc(204)] PcmOpen: plug:default,No such device or address

# (after running for a while)

CEF:Warning: libX11-xcb.so.1: cannot open shared object file: No such file or directory
CEF:Warning: loader_icd_scan: Failed loading library associated with ICD JSON /usr/lib/x86_64-linux-gnu/libvulkan_intel.so. Ignoring this JSON
CEF:Warning: libLLVM-17.so.1: cannot open shared object file: No such file or directory
CEF:Warning: loader_icd_scan: Failed loading library associated with ICD JSON /usr/lib/x86_64-linux-gnu/libvulkan_lvp.so. Ignoring this JSON
CEF:Warning: libLLVM-17.so.1: cannot open shared object file: No such file or directory
CEF:Warning: loader_icd_scan: Failed loading library associated with ICD JSON /usr/lib/x86_64-linux-gnu/libvulkan_radeon.so. Ignoring this JSON
CEF:Warning: libzstd.so.1: cannot open shared object file: No such file or directory
CEF:Warning: loader_icd_scan: Failed loading library associated with ICD JSON /usr/lib/x86_64-linux-gnu/libvulkan_virtio.so. Ignoring this JSON
CEF:Warning: vkCreateInstance: Found no drivers!
CEF:Warning: vkCreateInstance failed with VK_ERROR_INCOMPATIBLE_DRIVER
CEF:    at CheckVkSuccessImpl (../../third_party/dawn/src/dawn/native/vulkan/VulkanError.cpp:106)
CEF:
CEF:[2751991:2751991:0119/164219.507537:ERROR:gpu_blocklist.cc(71)] Unable to get gpu adapter
```

Also is there any information for building on darwin? The derivation is very Linux-centric with no guardrails on Linux only dependencies used (x11, alsa-lib, etc) as Grayjay advertises being cross-platform.

TL;DR
- Builds successfully.
- No sound plays from application.
- Various shared object file warnings related to graphics.

normalcea commented:
Grayjay version 3 has been superseded by Grayjay version 4. Plugins break if the application isn't using the latest upstream release so this PR will need to be updated with version 4.

lzszt commented:
is there anything preventing this from being merged?

normalcea commented:
Downloading content functions as expected, but exporting those files to your filesystem errors out. 

![Screenshot From 2025-02-12 11-34-45](https://github.com/user-attachments/assets/3924aafe-eae4-46de-8670-7614eea49a3a)

lzszt commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,153 @@
+  '';
+
+  postInstall = ''
+    chmod +x $out/lib/grayjay/cef/dotcefnative
```
```suggestion
    chmod +x $out/lib/grayjay/cef/dotcefnative
    chmod +x $out/lib/grayjay/ffmpeg
```

normalcea commented:
Also, this package should have a desktop file installed as well since it doesn't come with one by default. 

```diff
diff --git a/pkgs/by-name/gr/grayjay/package.nix b/pkgs/by-name/gr/grayjay/package.nix
index 6f590fefcc7d..08b831318edc 100644
--- a/pkgs/by-name/gr/grayjay/package.nix
+++ b/pkgs/by-name/gr/grayjay/package.nix
@@ -27,6 +27,8 @@
   libsecret,
   gitUpdater,
   autoPatchelfHook,
+  makeDesktopItem,
+  copyDesktopItems,
   libgcc,
   krb5,
 }:
@@ -78,6 +80,18 @@ buildDotnetModule {
 
   nativeBuildInputs = [
     autoPatchelfHook
+    copyDesktopItems
+  ];
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "Grayjay";
+      exec = "Grayjay";
+      icon = "grayjay";
+      comment = "Cross platform media application for streaming and downloading media";
+      desktopName = "Grayjay Desktop";
+      categories = [ "Network" ];
+    })
   ];
 
   projectFile = [
@@ -109,6 +123,8 @@ buildDotnetModule {
     rm $out/lib/grayjay/Portable
     ln -s /tmp/grayjay-launch $out/lib/grayjay/launch
     ln -s /tmp/grayjay-cef-launch $out/lib/grayjay/cef/launch
+    mkdir -p $out/share/icons/hicolor/scalable/apps
+    ln -s $out/lib/grayjay/grayjay.png $out/share/icons/hicolor/scalable/apps/grayjay.png
   '';
 
   makeWrapperArgs = [
```

sigboe commented:
@lzszt Are there any more binaries that need to be set to executable in this package?

@samfundev Sorry for being impatient, and ignorant, but are you able to update the Pull request? And are there more things needed for this to be merged? Should we start asking nicely for one of the adults to merge this?

samfundev commented:
> Sorry for being impatient, and ignorant, but are you able to update the Pull request? And are there more things needed for this to be merged? Should we start asking nicely for one of the adults to merge this?

No problem. I've updated the pull request. As far as I'm aware there isn't anything else that's needed for this to be merged.

lzszt commented:
> @lzszt Are there any more binaries that need to be set to executable in this package?

As far as I'm aware there aren't any other binaries which need to be executable.

normalcea reviewed (approved):
DLGTM

NobbZ commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

NobbZ commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>grayjay</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

NobbZ commented:
Is the failing aarch build expected? I have to be honest though, this was an emulated build via binfmt.

sigboe commented:
> Is the failing aarch build expected? I have to be honest though, this was an emulated build via binfmt.
@NobbZ I dont know the answer, but I do know that this program builds for MacOS with ARM and X86, so I do believe it should work on ARM linux

normalcea commented:
If it's unable to be built for arm platforms then the `meta.platforms` should be just shortened to `[ "x86_64-linux" ]` instead of `lib.platforms.unix` if this issue can't be resolved.

samfundev commented:
Grayjay's website doesn't list an arm build for Linux so I've updated the meta.platforms to just include x86.

jchv commented:
First of all, thank you for working on this.

Secondly, It seems like Grayjay desktop v5 came out 3 days ago. Hopefully the last time it updates before this PR is able to be merged...

Finally, sync is broken for one reason or another. Here's what happens when I try to start sync from Android Grayjay. When I scan the QR code on Android and try to link, each time I select the button to perform the link, the desktop application spits out this error:
```
SyncSocketSession:PerformVersionCheck 3
Failed to run as responder: System.Exception: Invalid version
   at Grayjay.ClientServer.Sync.Internal.SyncSocketSession.PerformVersionCheck()
   at Grayjay.ClientServer.Sync.Internal.SyncSocketSession.HandshakeAsResponder()
   at Grayjay.ClientServer.Sync.Internal.SyncSocketSession.<StartAsResponder>b__34_0()
Session closed
```
The Android version I'm using is version 285.

I'm skeptical it has anything to do with the NixOS derivation. I tried running both the native Linux GrayJay and the Windows version under Wine and sync similarly doesn't really seem to work, just kind of doing nothing, though I don't see the same log output. Curious behavior to be sure.

Basic application usage is working great, though.

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

jchv commented:
One more issue: looks like file dialogs are crashing right now. Probably need `wrapGappsHook3`/maybe `gsettings-desktop-schemas`.

Reproduce by going to Settings -> Import (lower left of modal) -> Import Grayjay Export
```
CEF:
CEF:(dotcefnative:105490): GLib-GIO-ERROR **: 22:47:49.043: No GSettings schemas are installed on the system
An exception occurred in the IPC: Unable to read beyond the end of the stream.    at System.IO.Stream.ReadAtLeastAsyncCore(Memory`1 buffer, Int32 minimumBytes, Boolean throwOnEndOfStream, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at DotCef.DotCefProcess.<Start>b__34_3()
Receive loop stopped.
WebSocketEndpoint:Removed disconnected client cda7de5d-777a-4b1a-9222-51396abbc905
StateCasting:Disposing.
An exception occurred while handling UDP result: 'The operation was canceled.':    at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Net.Sockets.SocketReceiveFromResult>.GetResult(Int16 token)
   at System.Net.Sockets.UdpClient.<ReceiveAsync>g__WaitAndWrap|62_0(ValueTask`1 task)
   at FUTO.MDNS.MDNSListener.ReceiveLoopAsync(UdpClient client, CancellationToken cancellationToken)
An exception occurred while handling UDP result: 'The operation was canceled.':    at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Net.Sockets.SocketReceiveFromResult>.GetResult(Int16 token)
   at System.Net.Sockets.UdpClient.<ReceiveAsync>g__WaitAndWrap|62_0(ValueTask`1 task)
   at FUTO.MDNS.MDNSListener.ReceiveLoopAsync(UdpClient client, CancellationToken cancellationToken)
An exception occurred while handling UDP result: 'The operation was canceled.':    at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Net.Sockets.SocketReceiveFromResult>.GetResult(Int16 token)
   at System.Net.Sockets.UdpClient.<ReceiveAsync>g__WaitAndWrap|62_0(ValueTask`1 task)
   at FUTO.MDNS.MDNSListener.ReceiveLoopAsync(UdpClient client, CancellationToken cancellationToken)
An exception occurred while handling UDP result: 'The operation was canceled.':    at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.CancellationToken.ThrowIfCancellationRequested()
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Net.Sockets.SocketReceiveFromResult>.GetResult(Int16 token)
   at System.Net.Sockets.UdpClient.<ReceiveAsync>g__WaitAndWrap|62_0(ValueTask`1 task)
   at FUTO.MDNS.MDNSListener.ReceiveLoopAsync(UdpClient client, CancellationToken cancellationToken)
T:StateSync server socket had an unexpected error.
System.Net.Sockets.SocketException (4): Interrupted system call
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName)
   at System.Net.Sockets.Socket.Accept()
   at System.Net.Sockets.TcpListener.AcceptTcpClient()
   at Grayjay.ClientServer.States.StateSync.<StartAsync>b__23_0()
```

jchv reviewed (approved):
Thank you for your patience. I only have one nit remaining, which is that I believe the last commit message should be:

    grayjay: init at 5

as per [pkgs/README.md#commit-conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions) and with the updated version. (Probably should also update the PR title, but that doesn't really matter as much.)

I am assuming by this point this has felt like a tiring experience and I am sorry to have to throw another nit at you, but the contributing guidelines are what they are. (And even if they weren't, obviously it would be confusing if it said 4 in the commit message.)

That said, I think this is good to go as soon as that's done.

Crossing my fingers.

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

normalcea commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2294

FliegendeWurst commented on pkgs/by-name/gr/grayjay/package.nix:
```diff
@@ -0,0 +1,184 @@
+    libsecret
+  ];
+
+  passthru.updateScript = gitUpdater { };
```
I don't think this update script will properly update the npm hash

Pandapip1 replied:
Yea, this should be replaced with nix-update-script.

samfundev replied:
I've updated this to use nix-update-script but I'm not 100% the updating is working correctly. The time it takes to build is slowing down my ability to test things, but I'll try to keep testing.

jchv commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 368427`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>grayjay</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

samfundev commented:
@luc-spec Sure, I've opened the issues.

sigboe commented:
@luc-spec I was unable to reproduce these findings, I detailed my steps in samfundev's issue tracker

https://github.com/samfundev/nixpkgs/issues/1

Masrkai commented:
https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2
https://git-lfs.github.com/spec/v1 is not a valid attribute name: prebuilt/.gitattributes:1
sha256:c1053553fbff49e3bc136efcf3612ab3639baec7914fa407472a8e74b440ff4d is not a valid attribute name: prebuilt/.gitattributes:2


when i tried to use this package, it just kept using more resources (internet)  and never actually installing, "just repeating that message for infinity"

Pandapip1 commented:
Interesting. For me it just took about an hour of it seemingly hanging and then it decided to resolve itself.

jchv commented:
The build is just slow due to the fetching phase. It's gotta pull a ~270 MiB Git repository, plus submodules, plus binaries in LFS, from a probably fairly under-resourced GitLab instance. To make matters worse, Grayjay is technically non-free even though source is available, so everyone will have to sit through the build, as opposed to the usual behavior where almost everyone will be pulling through Hydra.

Exactly how long it takes will be highly variable on your internet connection and CPU and the load of the GitLab instance.

There might be ways to improve on this. Grayjay has a GitHub mirror and I reckon GitHub will probably usually behave better, and we still have FOD hashes to verify that the results will be the same. That having been said, I guess GitHub LFS probably bills pretty horribly on bandwidth, so sending hordes of Nix users to use the GitHub mirror may not be great either; probably would be worth double-checking first.

Anyway, a lot of people get notified when you reply here, so I think if anyone wants to raise an issue about the build performance it would best be done in a separate, new issue.
